### DISCUSSION TITLE
Page rg results unless redirected

### DISCUSSION THREAD

**Discussion by braddunbar** [nushell/nushell#16665] [Q&A]
### Question

`rg` does [not include built in pager support](https://github.com/BurntSushi/ripgrep/issues/86). This means you have to feed it to one yourself.

```
$ rg needle | less
```

However, `rg` knows it's being fed into another command and removes all the pretty output, so you have to pass the `--pretty` flag.

```
$ rg --pretty needle | less
```

I'd like to use an alias for this, but then I can't pass more options or pipe the results to another command (e.g. `rg | xargs sd needle replacement`) because the output has ansi escapes in it.

So using zsh, I have the following snippet that redirects `rg` to a pager unless it's being fed to another command. 

```zsh
rg() {
  if [ -t 1 ]; then
    command rg --pretty "$@" | less
  else
    command rg "$@"
  fi
}
```

My best attempt at this in nushell is the following.

```nu
def rg --wrapped [...args] {
  if (is-terminal --stdout) {
    ^rg --pretty $args | less
  } else {
    ^rg $args
  }
}
```

But the `is-terminal` portion doesn't work because it's not equivalent to `test -t 1`. `is-terminal --stdout` is true regardless whether the command is part of a pipeline.

```
~: if (is-terminal --stdout) { "terminal attached" } else { "no terminal" } | cat                                 12/17/2023 07:57:27 PM
terminal attached
~: if (is-terminal --stdout) { "terminal attached" } else { "no terminal" }                                       12/17/2023 07:57:29 PM
terminal attached
```

`test` fails similarly.

```
~: if (test -t 1 | complete | get exit_code) == 1 { "terminal attached" } else { "no terminal" }                  12/17/2023 07:59:21 PM
terminal attached
~: if (test -t 1 | complete | get exit_code) == 1 { "terminal attached" } else { "no terminal" } | cat            12/17/2023 07:59:24 PM
terminal attached
```

How should I write this in nushell that would be equivalent?

**Comment by fdncred** [comment]
This wouldn't work because the syntax is wrong.
```nushell
def rg --wrapped [...args] {
  if (is-terminal --stdout) == 1 {
    ^rg --pretty $args | less
  } else {
    ^rg $args
  }
}
```
This should behave better.
```nushell
def rg --wrapped [...args] {
  if (is-terminal --stdout) == true {
    ^rg --pretty $args | less
  } else {
    ^rg $args
  }
}
```

It may not fix your problem but at least the syntax is right. I'm not sure how to make is-terminal return false. I haven't looked into the command too deeply.

**Comment by braddunbar** [comment]
Oh, thank you @fdncred! I mixed my `test` attempt with the `is-terminal` attempt. I've corrected it above now though! It doesn't seem to work with the correct syntax either though.

**Comment by fdncred** [comment]
We use [is-terminal](https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html) but I'm not sure it is working as expected. I'm guessing there is a difference between is-being-redirected and is-terminal or maybe just how nushell implements pipelines. Maybe I'm wrong here. We could have a bug. Feel free to show a program with rust's is-terminal working the way you excpect.

**Comment by fdncred** [comment]
How would you call your `rg` script above where it doesn't have the `--pretty` switch?

**Comment by braddunbar** [comment]
> I'm guessing there is a difference between is-being-redirected and is-terminal or maybe just how nushell implements pipelines.

This was my thought too! I'm glad to hear I'm not completely off base.

> We use [is-terminal](https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html) but I'm not sure it is working as expected.

It seems to be working as expected, given that `stdout` *is* connected to a terminal. It's just that in zsh, each command is expected to have a separate `stdout` but in nu they are not. Maybe we just need another way to find out if this is the last command in a pipeline…?

> How would you call your rg script above where it doesn't have the --pretty switch?

Do you mean when would I need the `else` clause? Generally to pipe the output to another command. For example, to replace some test in a particular set of files as follows…

```nu
rg -l needle | sd foo bar $in
```

That command currently fails for me because when run with `--pretty` it includes ansi escapes that `sd` can't handle.
